src/frontend/xgboost/xgboost_json.h

```diff
@@ -0,0 +1,376 @@
+    } else {
+      return false;
+    }
+  }
+  /* \brief pop handler off of delegator's stack, relinquishing parsing */
+  bool pop_handler();
+  /* \brief store current JSON key
+   * \param str the key to store
+   * \param length the length of the str char array
+   */
+  void set_cur_key(const char *str, std::size_t length);
+  /* \brief retrieve current JSON key */
+  const std::string &get_cur_key();
+  /* \brief check if current JSON key is indicated key
+   * \param query_key the value to compare against current JSON key
+   */
+  bool check_cur_key(const std::string &query_key);
+  /* \brief if current JSON key is the indicated string, assign value to output
+   * \param key the JSON key for this output
+   * \param value the value to be assigned
+   * \param output reference to object to which the value should be assigned
+   */
+  template <typename ValueType>
+  bool assign_value(const std::string &key,
+                    ValueType &&value,
+                    ValueType &output);
+  template <typename ValueType>
+  bool assign_value(const std::string &key,
+                    const ValueType &value,
+                    ValueType &output);
+
+ private:
+  /* \brief the delegator which delegated parsing responsibility to this handler */
+  std::weak_ptr<Delegator> delegator;
+  /* \brief the JSON key for the object currently being parsed */
+  std::string cur_key;
+};
+
+/*! \brief JSON handler that ignores all delegated input*/
+class IgnoreHandler : public BaseHandler {
+ public:
+  using BaseHandler::BaseHandler;
+  bool Null();
+  bool Bool(bool b);
+  bool Int(int i);
+  bool Uint(unsigned u);
+  bool Int64(int64_t i);
+  bool Uint64(uint64_t u);
+  bool Double(double d);
+  bool String(const char *str, std::size_t length, bool copy);
+  bool StartObject();
+  bool Key(const char *str, std::size_t length, bool copy);
+  bool StartArray();
+};
+
+/*! \brief base handler for updating some output object*/
+template <typename OutputType> class OutputHandler : public BaseHandler {
+ public:
+  /*! 
+   * \brief construct handler to be added to given delegator's stack
+   * \param parent_delegator pointer to Delegator for this handler
+   * \param output the object to be modified during parsing
+   */
+  OutputHandler(std::weak_ptr<Delegator> parent_delegator,
+                OutputType &output)
+      : BaseHandler{parent_delegator}, m_output{output} {};
+  OutputHandler(std::weak_ptr<Delegator> parent_delegator,
+                OutputType &&output) = delete;
+
+ protected:
+  /* \brief the output value constructed or modified during parsing */
+  OutputType &m_output;
+};
+
+/*! \brief handler for array of objects of given type*/
+template <typename ElemType, typename HandlerType = BaseHandler>
+class ArrayHandler : public OutputHandler<std::vector<ElemType>> {
+ public:
+  using OutputHandler<std::vector<ElemType>>::OutputHandler;
+  bool Bool(ElemType b) {
+    this->m_output.push_back(b);
+    return true;
+  }
```

Comment:
^^ @wphicks

Reply:
Yes, that's right. Only `bool Bool(bool)` will ever be instantiated from the template, not `bool Bool(AnyOtherType)`. I'll go ahead and make it more explicit, though, as you suggest.

Reply:
Ah, I'm beginning to remember why I went this direction to begin with. As written, the suggested change won't compile, but I'm assuming you meant something more along the lines of:

```
template <ArgType = ElemType>
typename std::enable_if<std::is_same<ArgType, bool>::value, bool>::type
Bool(ArgType b) override {
  this->m_output.push_back(b);
  return true;
}

template <ArgType = ElemType>
typename std::enable_if<std::is_same<ArgType, bool>::value, bool>::type
Bool(ArgType b) {
  return false;
}
```
This won't work either because a template cannot override a virtual function, so the base `Bool` is always called. The alternative is to create a `store_bool` function similar to what I did [here](https://github.com/dmlc/treelite/pull/202/files#diff-38fd03e89a1597a342c7600bd8ccede4R191-R207).

I guess I'd lean slightly toward leaving it as is and adding a comment than adding 2 additional methods, a layer of indirection, and more template metaprogramming, but I'm very happy to be overruled. It makes sense to me because I wrote it, but it's far more important that it makes sense to you and any other developers who will have to work with this later on.

What do you think, @hcho3?

Reply:
Got it. In that case, please add a comment that clarifies the intention of `bool Bool(ElemType b)`.